Assets/Mirage/Runtime/InterestManagement/InterestManager.cs

```diff
@@ -0,0 +1,286 @@
+        {
+            CheckGrowVisibilitySystem();
+
+            var stopWatch = Stopwatch.StartNew();
+
+            if (_visibilitySystems.Equals(system))
+            {
+                logger.LogWarning(
+                    "[InterestManager] - System already register to interest manager. Please check if this was correct.");
+
+                return;
+            }
+
+            if(logger.logEnabled)
+                logger.Log($"[Interest Manager] - Registering system {system} to our manager.");
+
+            for (int i = 0; i < _visibilitySystems.Length; i++)
+            {
+                if(!_visibilitySystems[i].Equals(default)) continue;
+
+                _visibilitySystems[i] = system;
+
+                stopWatch.Stop();
+
+                if (logger.logEnabled)
+                    logger.Log($"[Interest Manager] - RegisterVisibilitySystem Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+
+                return;
+            }
+
+            stopWatch.Stop();
+
+            if (logger.logEnabled)
+                logger.Log($"[Interest Manager] - RegisterVisibilitySystem Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+        }
+
+        /// <summary>
+        ///     Un-register a specific interest management system from the interest manager.
+        /// </summary>
+        /// <param name="system">The system we want to un-register from the interest manager.</param>
+        internal void UnRegisterVisibilitySystem(ref ObserverData system)
+        {
+            var stopWatch = Stopwatch.StartNew();
+
+            for (int i = 0; i < _visibilitySystems.Length; i++)
+            {
+                if (!_visibilitySystems[i].Equals(system)) continue;
+
+                _visibilitySystems[i] = default;
+
+                if (logger.logEnabled)
+                    logger.Log($"[Interest Manager] - Un-Registering system {system} from our manager.");
+
+                stopWatch.Stop();
+
+                if (logger.logEnabled)
+                    logger.Log($"[Interest Manager] - UnRegisterVisibilitySystem  Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+
+                return;
+            }
+
+            logger.LogWarning(
+                "[InterestManager] - Cannot find system in interest manager. Please check make sure it was registered.");
+
+            stopWatch.Stop();
+
+            if (logger.logEnabled)
+                logger.Log($"[Interest Manager] - UnRegisterVisibilitySystem Method Execution Time: {stopWatch.Elapsed.TotalMilliseconds} ms");
+        }
+
+        /// <summary>
+        ///     Find out all the players that can see an object
+        /// </summary>
+        /// <param name="identity">The identity of the object we want to check if player's can see it or not.</param>
+        /// <returns></returns>
+        internal HashSet<INetworkPlayer> Observers(NetworkIdentity identity)
```

Comment:
The way this looks atm it will be called once for every send call, which will likely become very expensive compared to a field that we use now

Reply:
We use a field in network identity. That field gets updated by current visibility systems through update method. Its almost exact same thing. This is now a central system versus network identity holding and updating info of visibility which it never should be doing. Open to suggestions but its almost identical to current system